### DISCUSSION TITLE
RITA and Security Onion 2.3.180

### DISCUSSION THREAD

**Discussion by ng1991** [Security-Onion-Solutions/securityonion#9346] [Unsupported Versions] [answered]
Hi,

I have started working with Security Onion and been requested to integrate RITA and although I know it isn't fully supported, I have followed the following two guides:

https://docs.securityonion.net/en/2.3/rita.html?highlight=rita
https://docs.securityonion.net/en/2.3/other-supported-logs.html?highlight=rita

These do not use the weslambert build (https://github.com/weslambert/securityonion-rita) is this the better option?

In terms of the current setup I have installed RITA on our forward nodes and I am using the below to create the CSV files:

rita import --rolling /nsm/zeek/logs/current $HOSTNAME
rita show-beacons $HOSTNAME > /nsm/rita/beacons.csv
rita show-exploded-dns $HOSTNAME > /nsm/rita/exploded-dns.csv
rita show-long-connections $HOSTNAME > /nsm/rita/long-connections.csv

If I look at the csv files created they are just a standard comma separated csv file. 

I have also enabled rita on the SO minion.sls and can see the configuration is added to the yml file after i restart the filebeat container:
![image](https://user-images.githubusercontent.com/115169428/206666910-6ed5fe7f-8c09-4f7e-89ae-0b843989c4e2.png)

From the log I am not seeing much bar the below however I saw this in another post and this wasnt the issue and it was actually a shard problem (https://github.com/Security-Onion-Solutions/securityonion/discussions/8352)

{"log.level":"error","@timestamp":"2022-12-07T21:33:22.372Z","log.logger":"input.filestream","log.origin":{"file.name":"filestream/prospector.go","file.line":294},"message":"Error while stopping harvester group: task failures\n\terror while adding new reader to the bookkeeper harvester is already running for file\n\terror while adding new reader to the bookkeeper harvester is already running for file\n\terror while adding new reader to the bookkeeper harvester is already running for file","service.name":"filebeat","id":"rita-dns","prospector":"file_prospector","ecs.version":"1.6.0"}

I am unsure at this point how to troubleshoot this further so any assistance anyone can provide would be appreciated as in theory I cant really see any obvious problems

**Comment by ng1991** [comment]
Just to add to the above I have also tried the git repo Wes provides and ran that on the manager, I did get the below error:

![image](https://user-images.githubusercontent.com/115169428/206718714-0e09f087-0c48-45d9-87a4-224196bc8780.png)

However after running "sudo salt-call state.apply rita saltenv=custom queue=True" on one of my forward/sensor nodes and then so-rita-start followed by so-rita-update the files have successfully created in the /nsm/rita directory:
 
![image](https://user-images.githubusercontent.com/115169428/206719770-8cd48964-74a9-4ffe-bfc3-23ee60b31fc3.png)

I still however dont see any logs and I have added the rita enable to the minion sls and can see the filebeat.yml has the inputs
![image](https://user-images.githubusercontent.com/115169428/206719278-5391ca29-0292-481d-be1a-65821610c7d8.png)

Anyone else facing similar or have any suggestions to where I am going wrong?

I know this isnt supported so really just looking for any help

Thanks

**Comment by ng1991** [comment]
Increased log level for filebeat to debug and can see that the files are getting picked up but the inputs, also increased log level on Elasticsearch and logstash to check on the manager and search nodes but didn't see anything that helped further debug issue

![image](https://user-images.githubusercontent.com/115169428/207036782-7055510d-b6d0-4d17-89ff-1cfee6bcecd2.png)

**Comment by weslambert** [comment]
Does a regular Saltstack highstate complete without issue?

if you check the Elasticsearch indices from the CLI (from a search node, or manager node if running a traditional Elastic cluster), do you see an index for RITA logs?

`so-elasticsearch-indices-list | grep rita`

Do you see any clues in `/var/log/logstash/logstash.log`?

**Reply by ng1991** [reply]
Hi Wes,

Really appreciate you getting back to me on this. 

That is one thing I had noticed which I hadn't mentioned is that the indices don't exist for RITA on my manager or search nodes.
![image](https://user-images.githubusercontent.com/115169428/207072488-e8aed4c2-6a81-4503-b91a-1a4f356aac5a.png)

I just ran salt-call state.highstate on my forwarder node and it completed ok
![image](https://user-images.githubusercontent.com/115169428/207073870-c35693c6-008b-420a-b143-38e64f46c953.png)

In terms of logstash the logs I have been looking at are in /opt/so/log/logstash/logstash.log and even when I amended from info to debug I couldn't see anything on the manager or the 2 search nodes we have related to rita

**Reply by ng1991** [reply]
@weslambert sorry to message again just wondering if you had any ideas on what else I could check, the filebeat log from above to me indicates it is being processed but the logstash logs on the manager and search nodes show nothing for rita which explains why the index isnt created on the search nodes as the data isn't passing

Manager logstash:

![image](https://user-images.githubusercontent.com/115169428/207606181-85b3475f-c4c1-4dea-911e-49ff287331ec.png)
![image](https://user-images.githubusercontent.com/115169428/207606404-0f709955-957b-4178-b17a-15318ea70510.png)

Search Node:
![image](https://user-images.githubusercontent.com/115169428/207606504-fd9203c7-c792-4cff-a969-d72d2215bec4.png)

only errors on search are for so-zeek  when it cant index a entry

**Reply by weslambert** [reply]
What happens if you add a new event to one of the log files, or copy it, delete the original, then add it again in the log directory? Are the logs being continually updated?

**Reply by ng1991** [reply]
Yes the logs are continually updating just nothing related to rita, if for example I restart the container I see logstash start and listen on port 5644 which filebeat on the sensors is outputting too.

The zeek logs are coming through from all the sensors it is just the new rita logs that never deliver.

I did play with changing the log level via api to the container also by checking "curl -XGET 'localhost:9600/_node/logging?pretty'' and changing the beats input to debug but still no extra information, all I get from logstash logs is the below on the manager
![image](https://user-images.githubusercontent.com/115169428/207714264-bb623953-68fc-4d0e-8e9b-d3e7113cbf84.png)

I see a bit more with errors on the search nodes but I don't know how to find out if the rita data is even getting past the manager logstash/redis

I am not sure how to troubleshoot the problem further as everything seems correct.

From filebeat I can see in debug mode the file is being processed, but cant get any information out of the logstash to show if its reaching it and not sure how too.
![image](https://user-images.githubusercontent.com/115169428/207714817-5ad5fc0a-373a-41a3-b7ac-1b00b2313ee2.png)

Is rita confirmed working in this version, it just feels potentially like I am hitting a bug as I have followed all the guides as discussed above and I am getting no errors.

**Reply by weslambert** [reply]
Another option is have a look at the pipeline stats for RITA. Do you see anything that indicates records are being processed when running the following command on the search nodes?

`for i in rita.beacon rita.connection rita.dns; do sudo so-elasticsearch-pipeline-stats $i; done`

**Reply by ng1991** [reply]
Thanks for getting back to me again.

Unfortunately that returns a count of 0 on the search nodes.

As a test on the forward node I copied the yml file to the local directory and hardcoded the below in so the only input is Rita and it should just dump to file which is working correctly:

filebeat.inputs:
#------------------------------ Log prospector --------------------------------
- type: filestream
  id: rita-beacon
  paths:
    - /nsm/rita/beacons.csv
  exclude_lines: ['^Score', '^Source', '^Domain', '^No results']
  fields:
    module: rita
    dataset: beacon
    category: network
  processors:
  - drop_fields:
      fields: ["source", "prospector", "input", "offset", "beat"]
  fields_under_root: true
  pipeline: "rita.beacon"
  index: "so-rita"

- type: filestream
  id: rita-connection
  paths:
    - /nsm/rita/long-connections.csv
    - /nsm/rita/open-connections.csv
  exclude_lines: ['^Source', '^No results']
  fields:
    module: rita
    dataset: connection
    category: network
  processors:
  - drop_fields:
      fields: ["source", "prospector", "input", "offset", "beat"]
  fields_under_root: true
  pipeline: "rita.connection"
  index: "so-rita"

- type: filestream
  id: rita-dns
  paths:
    - /nsm/rita/exploded-dns.csv
  exclude_lines: ['^Domain', '^No results']
  fields:
    module: rita
    dataset: dns
    category: network
  processors:
  - drop_fields:
      fields: ["source", "prospector", "input", "offset", "beat"]
  fields_under_root: true
  pipeline: "rita.dns"
  index: "so-rita"


#----------------------------- Elasticsearch/Logstash output ---------------------------------

output.file:
  path: "/tmp/filebeat"
  filename: filebeat


So I do actually see the file it is generated in the container tmp directory when I update the file with new data so the filebeat is working correctly and outputting the data

![image](https://user-images.githubusercontent.com/115169428/208454356-8d2dad0e-eb18-4cf7-ae9c-d8752e30624c.png)

**Reply by weslambert** [reply]
Try the following:

- Add `id => rita_output` to `opt/so/saltstack/default/salt/logstash/pipelines/config/so/9801_output_rita.conf`
- Restart Logstash
- Wait until you have confirmed new logs are being picked up by Filebeat
- On a search node, check if records have reached the Logstash RITA output configuration using `so-logstash-pipeline-stats search | grep -iC 10 rita`

**Reply by ng1991** [reply]
Hi Wes and again really appreciate the help you are providing.

I think the above may of highlighted the problem from the start, on my search nodes when i look in the /opt/so/conf/logstash/pipelines/search/ directory on my search nodes I cant see the rita output which I presume would explain why it never gets to elastic search. Below is a screenshot of the inputs and outputs I can see on the search nodes:
![image](https://user-images.githubusercontent.com/115169428/208943032-2487ec18-683d-43f5-acaf-2e565159519d.png)

I can see the rita file you mentioned though on the manager and I added the config you requested
![image](https://user-images.githubusercontent.com/115169428/208942957-9d5a4c3c-6ea3-4c3d-a460-84245738537a.png)

Any idea why the SALT would not be pushing this output?

**Reply by ng1991** [reply]
Think issue has been identified we have a unique need case and a copy of the search sls was made to add a custom output

I am now adding in the rita output as we still need the custom one, the issue was our side so sorry for the wasted time, do you know of a better way of making our custom output more permanent so if new outputs are added we don't miss them?

**Comment by ng1991** [comment, marked as answer]
Problem for this was caused by a custom modification made to logstash, this stopped new outputs from being published to the search nodes resulting in rita data not being processed and explaining why I saw no errors in the logs.

As we still needed the bespoke output we added the extra missing outputs into our custom search.sls located:

/opt/so/saltstack/local/pillar/logstash/search.sls

We added in the ones that were missing in our custom config from /opt/so/saltstack/default/pillar/logstash/search.sls

More detail can be seen in the chat.

Thank you @weslambert for you help in finding this issue
